File: code/dart/modular_api/lib/src/core/logger/logger.dart
import 'dart:convert';

import 'modular_logger.dart';
import 'log_sink.dart';

export 'modular_logger.dart' show LogLevel, ModularLogger;

/// Per-request logger that carries `traceId` and respects `logLevel` filtering.
///
/// Created by `LoggingMiddleware` for each incoming HTTP request and injected
/// into the UseCase via the `logger` property.
///
/// Accepts an optional [sink] for output — defaults to `stdout`.
/// In tests, pass a `StringBuffer` to capture output without side-effects.
/// Builds platform-specific correlation fields from the ids the framework resolved.
///
/// The framework emits open formats and nothing vendor-specific (roadmap invariant 7),
/// so a field like Google's `logging.googleapis.com/trace` — which needs a project id
/// the framework has no business knowing — is produced by the application:
///
/// ```dart
/// traceFieldFormatter: (traceId, spanId) => {
///   'logging.googleapis.com/trace': 'projects/my-project/traces/$traceId',
///   if (spanId != null) 'logging.googleapis.com/spanId': spanId,
/// }
/// ```
typedef TraceFieldFormatter = Map<String, Object> Function(
  String traceId,
  String? spanId,
);

class RequestScopedLogger implements ModularLogger {
  final String traceId;
  final LogLevel logLevel;
  final String serviceName;
  final StringSink _sink;

  /// The caller's `X-Request-ID`, when it sent one.
  ///
  /// Preserved beside [traceId] rather than promoted into it (runbook D6), so anyone
  /// correlating by the token they sent keeps working.
  final String? requestId;

  /// Builds platform correlation fields. `null` means none are emitted.
  final TraceFieldFormatter? traceFieldFormatter;

  /// The active server span's id, when tracing is configured.
  ///
  /// Known at construction because the tracing middleware is **outermost** (runbook
  /// D12 as reversed): the span already exists when `loggingMiddleware` creates this
  /// logger, so nothing needs mutating afterwards. An earlier design had logging
  /// outside tracing and had to attach this later, which is the scaffolding that
  /// reversal removed.
  final String? spanId;

  RequestScopedLogger({
    required this.traceId,
    required this.logLevel,
    required this.serviceName,
    this.spanId,
    this.requestId,
    this.traceFieldFormatter,
    StringSink? sink,
  }) : _sink = sink ?? defaultLogSink();

  // ─── Public API (8 RFC 5424 levels) ─────────────────────────────

  @override
  void emergency(String msg, {Map<String, dynamic>? fields}) =>
      _log(LogLevel.emergency, msg, fields: fields);

  @override
  void alert(String msg, {Map<String, dynamic>? fields}) =>
      _log(LogLevel.alert, msg, fields: fields);

  @override
  void critical(String msg, {Map<String, dynamic>? fields}) =>
      _log(LogLevel.critical, msg, fields: fields);

  @override
  void error(String msg, {Map<String, dynamic>? fields}) =>
      _log(LogLevel.error, msg, fields: fields);

  @override
  void warning(String msg, {Map<String, dynamic>? fields}) =>
      _log(LogLevel.warning, msg, fields: fields);

  @override
  void notice(String msg, {Map<String, dynamic>? fields}) =>
      _log(LogLevel.notice, msg, fields: fields);

  @override
  void info(String msg, {Map<String, dynamic>? fields}) =>
      _log(LogLevel.info, msg, fields: fields);

  @override
  void debug(String msg, {Map<String, dynamic>? fields}) =>
      _log(LogLevel.debug, msg, fields: fields);

  // ─── Framework-internal: request/response logging ───────────────

  /// Emits a "request received" log at `info` level.
  void logRequest({required String method, required String route}) {
    _log(
      LogLevel.info,
      'request received',
      extra: {'method': method, 'route': route},
    );
  }

  /// Emits a "request completed" log at the level determined by [statusCode].
  void logResponse({
    required String method,
    required String route,
    required int statusCode,
    required double durationMs,
    Map<String, dynamic>? extra,
  }) {
    _log(
      _levelForStatus(statusCode),
      'request completed',
      extra: {
        'method': method,
        'route': route,
        'status': statusCode,
        'duration_ms': durationMs,
        ...?extra,
      },
    );
  }

  /// Emits an "unhandled exception" log at `error` level.
  /// No stack trace, no exception message — by design (privacy/security).
  void logUnhandledException({
    required String route,
    required double durationMs,
  }) {
    _log(
      LogLevel.error,
      'unhandled exception',
      extra: {'route': route, 'status': 500},
    );
  }

  // ─── Internal helpers ───────────────────────────────────────────

  void _log(
    LogLevel level,
    String msg, {
    Map<String, dynamic>? fields,
    Map<String, dynamic>? extra,
  }) {
    // Filtering: only emit if the message level <= configured logLevel.
    if (level.value > logLevel.value) return;

    final entry = <String, dynamic>{
      'ts': DateTime.now().millisecondsSinceEpoch / 1000.0,
      'level': level.name,
      'severity': level.value,
      'msg': msg,
      'service': serviceName,
      'trace_id': traceId,
      if (spanId != null) 'span_id': spanId,
      if (requestId != null) 'request_id': requestId,
    };

    // Only when there is trace context to point at. A platform field naming a trace
    // that does not exist is worse than no field.
    final formatter = traceFieldFormatter;
    if (formatter != null && spanId != null) {
      entry.addAll(formatter(traceId, spanId));
    }

    if (extra != null) entry.addAll(extra);
    if (fields != null) entry['fields'] = fields;

    _sink.writeln(jsonEncode(entry));
  }

  /// Maps HTTP status code → RFC 5424 log level.
  static LogLevel _levelForStatus(int status) {
    if (status >= 500) return LogLevel.error;
    if (status >= 400) return LogLevel.warning;
    if (status >= 200 && status < 400) return LogLevel.info;
    // 1xx informational
    return LogLevel.notice;
  }
}

File: code/py/modular_api/src/modular_api/core/logger/logger.py
"""Structured JSON logger — RFC 5424 levels, request-scoped, zero dependencies.

Each log entry is a single JSON line with fixed fields:
``ts``, ``level``, ``severity``, ``msg``, ``service``, ``trace_id``.
Optional ``fields`` dict carries structured data from the caller.
"""

from __future__ import annotations

import json
import sys
import time
from enum import IntEnum
from typing import Callable, Protocol


class LogLevel(IntEnum):
    """RFC 5424 log levels in descending severity order.

    Filtering rule: if configured ``log_level = X``, only messages with
    ``value <= X`` are emitted.
    """

    emergency = 0  # system unusable
    alert = 1      # immediate action required
    critical = 2   # critical condition
    error = 3      # operation error, 5xx
    warning = 4    # abnormal condition, 4xx
    notice = 5     # normal but significant
    info = 6       # normal flow, 2xx/3xx
    debug = 7      # detailed diagnostics


class ModularLogger(Protocol):
    """Public logger interface exposed to UseCases.

    Each method corresponds to an RFC 5424 severity level.
    """

    @property
    def trace_id(self) -> str: ...
    def emergency(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def alert(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def critical(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def error(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def warning(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def notice(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def info(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def debug(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...


# Signature for the output sink — default writes to stdout.
WriteFn = Callable[[str], object]


def _default_write(line: str) -> None:
    sys.stdout.write(line + "\n")


#: Builds platform-specific correlation fields from the ids the framework resolved.
#:
#: The framework emits open formats and nothing vendor-specific (roadmap invariant 7), so
#: a field like Google's ``logging.googleapis.com/trace`` — which needs a project id the
#: framework has no business knowing — is produced by the application.
TraceFieldFormatter = Callable[[str, str | None], dict[str, object]]


class RequestScopedLogger:
    """Per-request logger carrying ``trace_id`` with ``log_level`` filtering.

    Created by the logging middleware for each HTTP request and injected
    into the UseCase via the ``logger`` property.

    Pass a custom ``write_fn`` for testability (capture output without I/O).
    """

    def __init__(
        self,
        *,
        trace_id: str,
        log_level: LogLevel,
        service_name: str,
        write_fn: WriteFn = _default_write,
        span_id: str | None = None,
        request_id: str | None = None,
        trace_field_formatter: TraceFieldFormatter | None = None,
    ) -> None:
        self._trace_id = trace_id
        self._log_level = log_level
        self._service_name = service_name
        self._write_fn = write_fn
        # Known at construction because the tracing middleware is outermost (runbook D12
        # as reversed): the span already exists when logging_middleware creates this
        # logger, so nothing needs mutating afterwards.
        self._span_id = span_id
        # The caller's X-Request-ID, preserved beside the trace id rather than promoted
        # into it (runbook D6).
        self._request_id = request_id
        self._trace_field_formatter = trace_field_formatter

    @property
    def trace_id(self) -> str:
        return self._trace_id

    # ── Public API (8 RFC 5424 levels) ──────────────────────────────

    def emergency(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.emergency, msg, fields=fields)

    def alert(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.alert, msg, fields=fields)

    def critical(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.critical, msg, fields=fields)

    def error(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.error, msg, fields=fields)

    def warning(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.warning, msg, fields=fields)

    def notice(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.notice, msg, fields=fields)

    def info(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.info, msg, fields=fields)

    def debug(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.debug, msg, fields=fields)

    # ── Framework-internal: request/response logging ────────────────

    def log_request(self, *, method: str, route: str) -> None:
        """Emit a 'request received' log at info level."""
        self._log(LogLevel.info, "request received", extra={"method": method, "route": route})

    def log_response(
        self,
        *,
        method: str,
        route: str,
        status_code: int,
        duration_ms: float,
        extra: dict[str, object] | None = None,
    ) -> None:
        """Emit a 'request completed' log at the level determined by status_code."""
        self._log(
            _level_for_status(status_code),
            "request completed",
            extra={
                "method": method,
                "route": route,
                "status": status_code,
                "duration_ms": duration_ms,
                **(extra or {}),
            },
        )

    def log_unhandled_exception(self, *, route: str, duration_ms: float) -> None:
        """Emit an 'unhandled exception' log at error level. No stack trace by design."""
        self._log(LogLevel.error, "unhandled exception", extra={"route": route, "status": 500})

    # ── Internal ────────────────────────────────────────────────────

    def _log(
        self,
        level: LogLevel,
        msg: str,
        *,
        fields: dict[str, object] | None = None,
        extra: dict[str, object] | None = None,
    ) -> None:
        # Filtering: only emit if message level <= configured log_level.
        if level.value > self._log_level.value:
            return

        entry: dict[str, object] = {
            "ts": time.time(),
            "level": level.name,
            "severity": level.value,
            "msg": msg,
            "service": self._service_name,
            "trace_id": self._trace_id,
        }

        if self._span_id is not None:
            entry["span_id"] = self._span_id
        if self._request_id is not None:
            entry["request_id"] = self._request_id

        # Only when there is trace context to point at. A platform field naming a trace
        # that does not exist is worse than no field.
        if self._trace_field_formatter is not None and self._span_id is not None:
            entry.update(self._trace_field_formatter(self._trace_id, self._span_id))

        if extra is not None:
            entry.update(extra)
        if fields is not None:
            entry["fields"] = fields

        self._write_fn(json.dumps(entry))


def _level_for_status(status: int) -> LogLevel:
    """Map HTTP status code to RFC 5424 log level."""
    if status >= 500:
        return LogLevel.error
    if status >= 400:
        return LogLevel.warning
    if status >= 200:
        return LogLevel.info
    return LogLevel.notice  # 1xx
